rescript/get_data.py

```diff
@@ -12,13 +12,132 @@
+        
+    response = requests.get(url, stream=True)
+    if response.status_code != 200:
+        raise ValueError("Failed to download the file from " + url)
+    
+    tar_file_path = os.path.join(download_path, 'unitefile.tar.gz')
+    with open(tar_file_path, 'wb') as f:
+        f.write(response.content)
+    
+    # Extract only the 'developer' subdirectory
+    with tarfile.open(tar_file_path, 'r:gz') as tar:
+        # Ensure that 'developer' exists in the tar file
+        members = [member for member in tar.getmembers() if member.name.startswith('developer')]
+        if not members:
+            raise ValueError("No 'developer' subdirectory found in the .tar.gz file.")
+        
+        for member in members:
+            member.name = os.path.basename(member.name) # Strip the 'developer' prefix
+            tar.extract(member, path=download_path)
+    
+    return download_path
+
+# Test it by downloading this file
+# _unite_download_targz('https://files.plutof.ut.ee/public/orig/59/12/591225E8985EFC44B595C79AF5F467421B4D9A95093A0811B13CB4CC13A6DA46.tgz', tmp_dir)
+
+# import as artifacts
+# results[name] = qiime2.Artifact.import_data(dtype, destination)
+
+def get_unite_data(version, taxon_group, singletons=False):
+    url = _unite_get_url(version, taxon_group, singletons)
+    results = {'sequences': [], 'taxonomy': []}
+    
+    # with tempfile.TemporaryDirectory() as tmp_dir:
+    tmp_dir = tempfile.mkdtemp()
+    
+    print('Temporary directory:', tmp_dir)
+    _unite_download_targz(url, download_path=tmp_dir)
+    
+    for root, dirs, files in os.walk(tmp_dir):
+        for file in files:
+            print(results)
+            if file.endswith('.fasta'):
+                fasta_file_name = os.path.join(root, file)
+                print('found fasta: ' + fasta_file_name)
+                with open(fasta_file_name, 'r') as fasta_file:
+                    # Read the content of the file and append it as a Python object
+                    fasta_content = fasta_file.read()
+                    results['sequences'].append(qiime2.Artifact.import_data('FeatureData[RNASequence]', fasta_content))
+            elif file.endswith('.txt'):
+                txt_file_name = os.path.join(root, file)
+                print('found txt: ' + txt_file_name)
+                results['taxonomy'].append(qiime2.Artifact.import_data('FeatureData[Taxonomy]', txt_file_name))
+    return results
+
+get_unite_data(version='9.0', taxon_group='fungi')
+
+
+# How do I import data?
+with open("/tmp/tmpzprzopiu/sh_refs_qiime_ver9_97_25.07.2023_dev.fasta", 'r') as fasta_file:
+    qiime2.Artifact.import_data('FeatureData[RNASequence]', fasta_file)
+
+with open("/tmp/tmpzprzopiu/sh_refs_qiime_ver9_97_25.07.2023_dev.fasta", 'r') as fasta_file:
+    # Read the content of the file and append it as a Python object
+    fasta_content = fasta_file.read()
+    qiime2.Artifact.import_data('FeatureData[RNASequence]', fasta_content)
+
```

Comment:
Neither of these work. How do I import data?

Reply:
you don't need to import the data as Q2 artifacts unless if this action is being set up as a pipeline (as the other get-data actions have been so far), you would just output something like a `DNAFASTAFormat` object containing the sequences and QIIME 2 will do the rest.

But if you want to set this up as a Q2 pipeline, I _think_ that this would work for importing:
```
with open("/tmp/tmpzprzopiu/sh_refs_qiime_ver9_97_25.07.2023_dev.fasta", 'r') as fasta_file:
    qiime2.Artifact.import_data('FeatureData[RNASequence]', str(fasta_file))
```

Reply:
```python
fasta_file = open("/tmp/tmpduape95o/sh_refs_qiime_ver9_97_25.07.2023_dev.fasta", 'r')

qiime2.Artifact.import_data('FeatureData[Sequence]', fasta_file)
qiime2.Artifact.import_data('FeatureData[Sequence]', str(fasta_file))
qiime2.Artifact.import_data('FeatureData[RNASequence]', fasta_file)
qiime2.Artifact.import_data('FeatureData[RNASequence]', str(fasta_file))
qiime2.Artifact.import_data('FeatureData[DNASequence]', fasta_file)
qiime2.Artifact.import_data('FeatureData[DNASequence]', str(fasta_file))
```

none of these "have a compatible directory format."

Reply:
>this action is being set up as a pipeline (as the other get-data actions have been so far)

I was thinking to mirror functionality of get_silva*. What do you recommend?

Reply:
get-silva-data is a pipeline because it calls other actions for parsing the silva file formats/taxonomy. So get-unite-data could be simpler... I would just write these to a DNAFASTAFormat object and let Q2 do the rest. You can see an example here: https://github.com/bokulich-lab/RESCRIPt/blob/master/rescript/degap.py

Reply:
Look to the [GTDB code](https://github.com/bokulich-lab/RESCRIPt/blob/master/rescript/get_gtdb.py). The data was already in a QIIME compatible format, except for being lower case. Some examples for the imports:

-  [merging](https://github.com/bokulich-lab/RESCRIPt/blob/946728467532f066c8bb750c3d8d1b8af02773b9/rescript/get_gtdb.py#L46)
-  [tar](https://github.com/bokulich-lab/RESCRIPt/blob/946728467532f066c8bb750c3d8d1b8af02773b9/rescript/get_gtdb.py#L129C2-L129C2)
- [gzip](https://github.com/bokulich-lab/RESCRIPt/blob/946728467532f066c8bb750c3d8d1b8af02773b9/rescript/get_gtdb.py#L136)